ramls/holdingsrecord-with-id.json

```diff
@@ -0,0 +1,282 @@
+{
```

Comment:
No, because the `id` property is optional in `holdingsrecord.json` and required in `holdingsrecord-with-id.json`: https://github.com/folio-org/mod-inventory-storage/blob/6cf0931ca346523ca30f0f39a90161b78df9b0ad/src/test/java/org/folio/rest/unit/SchemaTest.java#L24

Reply:
As far as I understand, these endpoints will be used for instances/holdings/items upload, not for updates. Why do we need ids to be required?

Reply:
Personally, I'd prefer to have a java based validation for the id then, instead of maintaining two separate files of the schemes (considering that these schemes are changed frequently).

What do you think @marcjohnson-kint, @mreno-EBSCO?

Reply:
This is a long running challenge with using the same schema for both read and write representations of the same record, it leads to compromises like this.

The `id` property is only optional in the record schema because it can be specified by the client during creation (using POST). In all other circumstances it should be effectively required.

If we decide to have two schema, I think we need to figure out how to help maintainers know they need to change both schema together.

I personally think we could do the check after the records have been received and issue a 422 if the ID isn't present. We do this for other conditional validation that cannot be expressed in the schema.

Which trade off do we want to make?

If we do decide to use two schema, I'd prefer we investigate adopting read and write schema (or similar, because create is slightly different to update) rather than two variants of the schema whose usage is less clear. I'm aware this isn't ideal.

Including @skoczko in this conversation as well.

Reply:
Since we use the schema for API documentation, I think that has to be accurate as possible. Indicating that an ID is not required in the schema, but in the code making required in certain cases makes the APIs harder to use, IMHO. There are several cases where this occurs in existing modules and it is frustrating.

Reply:
The id is required because this API does not return anything and therefore cannot report the generated id.
The client that calls this API needs the instance id to put it into the holding.instanceId field. The same for item.holdingsRecordId.
If the insert fails the id allows to easily track which records caused the failure.

Reply:
Having two schema that differ by id requirement only has several advantages:
* RMB can do the validation. No need to split the validation (schema and java).
* The schema shown in the API documentation https://dev.folio.org/reference/api/ is correct, one can rely on the list of required fields.
* Maintainers are forced to update both schemas (with and without id) at the same time because the schemasWithIdAreInSync unit test will fail if the files differ: https://github.com/folio-org/mod-inventory-storage/blob/938fcecd8ee96764704320f496854102a653e570/src/test/java/org/folio/rest/unit/SchemaTest.java#L27

Reply:
Thanks for explaining it, @julianladisch.

Reply:
I also think it's horrible to have two schemas and is in no way convinced about that being elengant because some module has a unit test for some schema. We started with DB generated UUID, then Java based UUID. The final stage would be client side generated in which case we'd have one schem only - with a required UUID. That would be a major change, but nutural step forward IMHO.

Reply:
How do we come to an agreement on this topic for this change?

I agree that having more descriptive schema is beneficial.

At the moment, that would require having (at least) two different schema. 

Is the improved description of the `id` property valuable enough to introduce that extra maintenance (inc. the tests)?

I think it is worth considering @adamdickmeiss suggestion for FOLIO standards, as it would remove the challenge around this particular property.

In general, I think there will often be a gap (conditional validation, external state etc) between what can be expressed in a schema and what the server considers valid, so clients will likely always need to be prepared for records that are valid against the schema to be rejected.

Reply:
If we were to introduce two schema, I’d prefer we made the special case (single record creation, POST and PUT) the explicit second schema and use the other schema for all other situations (assuming that is the distinction).